src/decoder/mod.rs

```diff
@@ -1455,60 +1218,159 @@ impl<R: Read + Seek> Decoder<R> {
+            tile_width,
+            byte_order,
+            tile,
+        )?;
 
         self.current_chunk += 1;
 
         Ok(result)
     }
 
-    fn read_tiled_image(&mut self) -> TiffResult<DecodingResult> {
+    fn read_chunked_image(&mut self) -> TiffResult<DecodingResult> {
         let width = usize::try_from(self.image().width)?;
-        let mut result = self.result_buffer(width, usize::try_from(self.image().height)?)?;
+        let height = usize::try_from(self.image().height)?;
+        let mut result = self.result_buffer(width, height)?;
+        if width == 0 || height == 0 {
+            return Ok(result);
+        }
 
-        for tile in 0..self.image().chunk_offsets.len() {
-            let buffer_offset = self
-                .image()
-                .tile_attributes
-                .as_ref()
-                .unwrap()
-                .get_offset(tile);
-            self.read_tile_to_buffer(&mut result.as_buffer(buffer_offset), tile, width)?;
+        let chunk_dimensions = self.image().chunk_dimensions()?;
+        let chunk_dimensions = (
+            chunk_dimensions.0.min(width),
+            chunk_dimensions.1.min(height),
+        );
+        if chunk_dimensions.0 == 0 || chunk_dimensions.1 == 0 {
+            return Err(TiffError::FormatError(
+                TiffFormatError::InconsistentSizesEncountered,
+            ));
         }
 
-        Ok(result)
-    }
+        let samples = self.image().bits_per_sample.len();
+        if samples == 0 {
+            return Err(TiffError::FormatError(
+                TiffFormatError::InconsistentSizesEncountered,
+            ));
+        }
 
-    fn read_stripped_image(&mut self) -> TiffResult<DecodingResult> {
-        let rows_per_strip =
-            usize::try_from(self.image().strip_decoder.as_ref().unwrap().rows_per_strip)?;
+        let chunks_across = (width - 1) / chunk_dimensions.0 + 1;
+        let strip_samples = width * chunk_dimensions.1 * samples;
 
-        let samples_per_strip = match usize::try_from(self.image().width)?
-            .checked_mul(rows_per_strip)
-            .and_then(|x| x.checked_mul(self.image().bits_per_sample.len()))
+        // Parallel decode. Requires the "rayon" feature, the presence of multiple vertical chunks
+        // in the image, and that all chunks are at most half the limit for intermediate buffer size.
+        #[cfg(feature = "rayon")]
+        if height > chunk_dimensions.1
+            && self
+                .image()
+                .chunk_bytes
+                .iter()
+                .all(|&b| b <= self.limits.intermediate_buffer_size as u64 / 2)
         {
-            Some(s) => s,
-            None => return Err(TiffError::LimitsExceeded),
-        };
+            let condvar = std::sync::Condvar::new();
+            let bytes_allocated = std::sync::Mutex::new(Ok(0));
+            for x in 0..chunks_across {
+                rayon::in_place_scope_fifo(|s| -> TiffResult<()> {
```

Comment:
We're already starting a new scope for each x coordinate, can't we ensure the allocation requirement by ensuring that at most `n` tasks run in parallel, where `n` is `intermediate_buffer_size / chunk_bytes.iter().max()`. Or an adaptive version instead.

I'd also be more comfortable if tasks were generated as an iterator instead of this somewhat complex case distinction of different loops, if that makes sense. In particular this makes it easier to transition to using some of the rayon parallel iterators in the future?

Reply:
The two loops are needed to work around a borrow checker issue for tiled images (for stripped images the outer loop executes exactly one iteration).

It may be possible to replace the inner loop with an iterator that generate the tasks. However, I think it would be tricky to implement that while still doing file reads and chunk expands in parallel